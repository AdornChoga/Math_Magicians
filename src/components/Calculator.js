import '../stylesheets/Calculator.css';
import React, { useState } from 'react';
import calculate from '../logic/calculate';
import Button from './Button';

// eslint-disable-next-line react/prefer-stateless-function
const Calculator = () => {
  const [state, setState] = useState({ total: 0, next: null, operation: null });
  const { total, next, operation } = state;

  const handleClicks = (e) => {
    const display = document.querySelector('#number-input');
    const expression = document.querySelector('#expression');
    const btnName = e.target.textContent;
    const updatedState = calculate(state, btnName);
    setState(updatedState);
    if (!updatedState.total && !updatedState.operation) {
      if (btnName === '0' && display.innerHTML === '0') {
        display.innerHTML = btnName;
        return;
      }
      if (btnName === '0') {
        display.innerHTML += btnName;
      }
    }
    if (Number(btnName) || btnName === '.' || btnName === '+/-') {
      if (!total && !next) {
        display.innerHTML = btnName;
      } else if (!updatedState.total && !updatedState.operation) {
        display.innerHTML = updatedState.next;
      } else {
        display.innerHTML += btnName;
      }
    }
    if (btnName === '+/-' && updatedState.total && !updatedState.next) {
      display.innerHTML = updatedState.total;
    } else if (btnName === '+/-' && updatedState.total && updatedState.next) {
      display.innerHTML = updatedState.next;
    }
    if (Number(btnName) && total && !next && !operation) {
      expression.innerHTML = '';
      display.innerHTML = btnName;
    }
    if (!updatedState.total && !updatedState.next && !updatedState.operation) {
      expression.innerHTML = '';
      display.innerHTML = '0';
    }
    if (updatedState.total && updatedState.operation) {
      expression.innerHTML = `${updatedState.total} ${updatedState.operation}`;
      if (Number(btnName)) {
        display.innerHTML = '';
        display.innerHTML += updatedState.next;
      }
    }
    if (updatedState.operation && !updatedState.next && !updatedState.total) {
      setState({ total: '0' });
      expression.innerHTML = `0 ${updatedState.operation}`;
      if (Number(btnName)) {
        display.innerHTML = '';
        display.innerHTML += btnName;
      }
    }
    if (e.target.textContent === '=') {
      if (updatedState.total !== undefined) {
        display.innerHTML = updatedState.total;
      }
      if (total && next && operation) {
        expression.innerHTML = `${total} ${operation} ${next} =`;
      }
    }
  };
  return (
    <ul className="grid-container">
      <li id="display">
        <span id="expression" />
        <span id="number-input">0</span>
      </li>
      <Button buttonValue="AC" eventHandler={handleClicks} />
      <Button buttonValue="+/-" eventHandler={handleClicks} />
      <Button buttonValue="%" eventHandler={handleClicks} />
      <Button buttonValue="รท" eventHandler={handleClicks} />
      <Button buttonValue="7" eventHandler={handleClicks} />
      <Button buttonValue="8" eventHandler={handleClicks} />
      <Button buttonValue="9" eventHandler={handleClicks} />
      <Button buttonValue="x" eventHandler={handleClicks} />
      <Button buttonValue="4" eventHandler={handleClicks} />
      <Button buttonValue="5" eventHandler={handleClicks} />
      <Button buttonValue="6" eventHandler={handleClicks} />
      <Button buttonValue="-" eventHandler={handleClicks} />
      <Button buttonValue="1" eventHandler={handleClicks} />
      <Button buttonValue="2" eventHandler={handleClicks} />
      <Button buttonValue="3" eventHandler={handleClicks} />
      <Button buttonValue="+" eventHandler={handleClicks} />
      <Button buttonValue="0" eventHandler={handleClicks} />
      <Button buttonValue="." eventHandler={handleClicks} />
      <Button buttonValue="=" eventHandler={handleClicks} />
    </ul>
  );
};

export default Calculator;
