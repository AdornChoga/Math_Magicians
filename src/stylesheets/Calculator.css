* { box-sizing: border-box; }

html {
  font-size: 62.5%;
}

body {
  font-family: sans-serif;
}

.grid-container {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr 1fr 1fr;
  max-width: 40rem;
  margin: 0 auto;
  padding: 0.5rem;
  list-style: none;
}

.grid-container li {
  border: solid #d7d7d7;
  padding: 0;
  text-align: center;
  font-size: 1.5rem;
}

.grid-container button {
  display: inline;
  padding: 2.2rem 1rem;
  border: none;
  width: 100%;
  background-color: #e0e0e0;
}

.grid-container button:hover {
  background-color: #f8f7f7;
}

.grid-container button:active {
  background-color: #cecaca;
}

.grid-container li:nth-child(4n + 1) button,
.grid-container li:last-child button {
  background-color: #f5913e;
}

.grid-container li:nth-child(4n + 1) button:hover,
.grid-container li:last-child button:hover {
  background-color: #f1af79;
}

.grid-container li:nth-child(4n + 1) button:active,
.grid-container li:last-child button:active {
  background-color: #f77e1b;
}

#display {
  grid-column: 1/5;
  background: #858694;
  text-align: right;
  color: white;
  font-size: 3rem;
  padding: 0.9rem 0;
}

#zero {
  grid-column: 1/3;
}
