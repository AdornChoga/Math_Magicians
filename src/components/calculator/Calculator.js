import './Calculator.css';
import React, { useState } from 'react';
import calculate from '../../logic/calculator/calculate';
import Button from '../button/Button';

// eslint-disable-next-line react/prefer-stateless-function
const Calculator = () => {
  const [state, setState] = useState({ total: 0, next: null, operation: null });
  const { total, next, operation } = state;

  const handleClicks = (e) => {
    const display = document.querySelector('#number-input');
    const expression = document.querySelector('#expression');
    const btnName = e.target.textContent;
    const updatedState = calculate(state, btnName);
    setState(updatedState);
    if (!updatedState.total && !updatedState.operation) {
      if (btnName === '0' && display.innerHTML === '0') {
        display.innerHTML = btnName;
        return;
      }
      if (btnName === '0') {
        display.innerHTML += btnName;
      }
    }
    if (Number(btnName) || btnName === '.' || btnName === '+/-' || btnName === '0') {
      if (!total && !next) {
        display.innerHTML = btnName;
      } else {
        display.innerHTML = updatedState.next;
      }
    }
    if (btnName === '+/-' && updatedState.total && !updatedState.next) {
      display.innerHTML = updatedState.total;
    } else if (btnName === '+/-' && updatedState.total && updatedState.next) {
      display.innerHTML = updatedState.next;
    }
    if (Number(btnName) && total && !next && !operation) {
      expression.innerHTML = '';
      display.innerHTML = btnName;
    }
    if (!updatedState.total && !updatedState.next && !updatedState.operation) {
      expression.innerHTML = '';
      display.innerHTML = '0';
    }
    if (updatedState.total && updatedState.operation) {
      expression.innerHTML = `${updatedState.total} ${updatedState.operation}`;
      if (Number(btnName)) {
        display.innerHTML = '';
        display.innerHTML += updatedState.next;
      }
    }
    if (updatedState.operation && !updatedState.next && !updatedState.total) {
      setState({ total: '0' });
      expression.innerHTML = `0 ${updatedState.operation}`;
      if (Number(btnName)) {
        display.innerHTML = '';
        display.innerHTML += btnName;
      }
    }
    if (e.target.textContent === '=') {
      if (updatedState.total !== undefined) {
        display.innerHTML = updatedState.total;
      }
      if (total && next && operation) {
        expression.innerHTML = `${total} ${operation} ${next} =`;
      }
    }
  };
  return (
    <div className="calc-page-container" data-testid="calc-page-cont">
      <h1>Let &apos;s do some Math!</h1>
      <ul className="grid-container">
        <li id="display">
          <span id="expression" />
          <span id="number-input">0</span>
        </li>
        <Button buttonValue="AC" eventHandler={handleClicks} id="clear" />
        <Button buttonValue="+/-" eventHandler={handleClicks} id="negate" />
        <Button buttonValue="%" eventHandler={handleClicks} id="percent" />
        <Button buttonValue="รท" eventHandler={handleClicks} id="divide" />
        <Button buttonValue="7" eventHandler={handleClicks} id="seven" />
        <Button buttonValue="8" eventHandler={handleClicks} id="eight" />
        <Button buttonValue="9" eventHandler={handleClicks} id="nine" />
        <Button buttonValue="x" eventHandler={handleClicks} id="multiply" />
        <Button buttonValue="4" eventHandler={handleClicks} id="four" />
        <Button buttonValue="5" eventHandler={handleClicks} id="five" />
        <Button buttonValue="6" eventHandler={handleClicks} id="six" />
        <Button buttonValue="-" eventHandler={handleClicks} id="minus" />
        <Button buttonValue="1" eventHandler={handleClicks} id="one" />
        <Button buttonValue="2" eventHandler={handleClicks} id="tow" />
        <Button buttonValue="3" eventHandler={handleClicks} id="three" />
        <Button buttonValue="+" eventHandler={handleClicks} id="plus" />
        <Button buttonValue="0" eventHandler={handleClicks} id="zero" />
        <Button buttonValue="." eventHandler={handleClicks} id="point" />
        <Button buttonValue="=" eventHandler={handleClicks} id="equal" />
      </ul>
    </div>

  );
};

export default Calculator;
