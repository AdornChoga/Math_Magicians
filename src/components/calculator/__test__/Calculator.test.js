import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../Calculator';

afterEach(cleanup);
describe('Tests for rendering and snapshot matching', () => {
  test('does calculator output a button component', () => {
    render(<Calculator />);
    const container = screen.getByTestId('calc-page-cont');
    expect(container).toBeInTheDocument();
  });

  test('is snapshot of calculator component created and matched', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
