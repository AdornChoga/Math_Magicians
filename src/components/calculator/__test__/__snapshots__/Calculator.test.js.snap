// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for rendering and snapshot matching is snapshot of calculator component created and matched 1`] = `
<div
  className="calc-page-container"
  data-testid="calc-page-cont"
>
  <h1>
    Let 's do some Math!
  </h1>
  <ul
    className="grid-container"
  >
    <li
      id="display"
    >
      <span
        id="expression"
      />
      <span
        id="number-input"
      >
        0
      </span>
    </li>
    <li>
      <button
        data-testid="clear"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
    </li>
    <li>
      <button
        data-testid="negate"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
    </li>
    <li>
      <button
        data-testid="percent"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
    </li>
    <li>
      <button
        data-testid="divide"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </li>
    <li>
      <button
        data-testid="seven"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
    </li>
    <li>
      <button
        data-testid="eight"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
    </li>
    <li>
      <button
        data-testid="nine"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
    </li>
    <li>
      <button
        data-testid="multiply"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </li>
    <li>
      <button
        data-testid="four"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </li>
    <li>
      <button
        data-testid="five"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </li>
    <li>
      <button
        data-testid="six"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
    </li>
    <li>
      <button
        data-testid="minus"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </li>
    <li>
      <button
        data-testid="one"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li>
      <button
        data-testid="tow"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li>
      <button
        data-testid="three"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li>
      <button
        data-testid="plus"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </li>
    <li>
      <button
        data-testid="zero"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
    </li>
    <li>
      <button
        data-testid="point"
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </li>
    <li>
      <button
        data-testid="equal"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </li>
  </ul>
</div>
`;
